File: bongf/java/src/dongbinbook/ch18_graph_questions/Q44_PlanetTunnel.java
package dongbinbook.ch18_graph_questions;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public class Q44_PlanetTunnel {
    public static int[] parent;
    public static void main(String[] args) throws IOException {
        System.out.println(solution());
    }

    public static int solution() throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        parent = new int[n];
        Planet[] x = new Planet[n];
        Planet[] y = new Planet[n];
        Planet[] z = new Planet[n];

        for (int i = 0; i < n; i++) {
            parent[i] = i;
            Integer[] input = Arrays.stream(br.readLine().split(" ")).map(Integer::parseInt).toArray(Integer[]::new);
            x[i] = new Planet(i, input[0]);
            y[i] = new Planet(i, input[1]);
            z[i] = new Planet(i, input[2]);
        }

        Arrays.sort(x);
        Arrays.sort(y);
        Arrays.sort(z);

        List<Path> paths = new ArrayList<>();
        for (int i = 0; i < n-1; i++) {
            paths.add(new Path(x[i+1].position - x[i].position, x[i].index, x[i+1].index));
            paths.add(new Path(y[i+1].position - y[i].position, y[i].index, y[i+1].index));
            paths.add(new Path(z[i+1].position - z[i].position, z[i].index, z[i+1].index));
        }

        Collections.sort(paths);
        int result = 0;
        for (Path path : paths) {
            if( findP(path.x) != findP(path.y)) {
                union(path.x, path.y);
                result += path.distance;
            }
        }
        return result;
    }

    public static int findP(int x) {
        if(parent[x] != x) {
            parent[x] = findP(parent[x]);
        }
        return parent[x];
    }

    public static void union(int x, int y) {
        x = findP(x);
        y = findP(y);
        if( x < y ) {
            parent[x] = y;
            return;
        }
        parent[y] = x;
    }
}

class Planet implements Comparable<Planet> {
    int index;
    int position;

    public Planet(int index, int position) {
        this.index = index;
        this.position = position;
    }

    @Override
    public int compareTo(Planet other) {
        return this.position - other.position;
    }
}

class Path implements Comparable<Path> {
    int distance;
    int x;
    int y;

    public Path(int distance, int x, int y) {
        this.distance = distance;
        this.x = x;
        this.y = y;
    }

    @Override
    public int compareTo(Path path) {
        return this.distance - path.distance;
    }
}

File: bongf/python/dongbinbook/ch18_graph_questions/q44_planet_tunnel.py
import sys
input = sys.stdin.readline
# https://www.acmicpc.net/problem/2887

def find_p(parents, x):
    if(parents[x] != x):
        parents[x] = find_p(parents, parents[x])
    return parents[x]

def union(parents, x, y):
    x = find_p(parents, x)
    y = find_p(parents, y)
    if x < y :
        parents[y] = x
    else :
        parents[x] = y

def dongbin():
    n = int(input())
    edges = []
    parent = [0] * n
    result = 0 

    x = []
    y = []
    z = []

    for i in range(n):
        x1, y1, z1 = map(int, input().split(" "))
        parent[i] = i
        x.append((x1, i))
        y.append((y1, i))
        z.append((z1, i))

    x.sort()
    y.sort()
    z.sort()

    for i in range(n-1):
        edges.append((x[i+1][0] - x[i][0], x[i][1], x[i+1][1]))
        edges.append((y[i+1][0] - y[i][0], y[i][1], y[i+1][1]))
        edges.append((z[i+1][0] - z[i][0], z[i][1], z[i+1][1]))
    
    edges.sort()
    for edge in edges:
        cost, a, b = edge
        if find_p(parent, a) != find_p(parent, b):
            union(parent, a, b)
            result += cost
    return result

print(dongbin())


# 메모리 초과 
def solution_fail():
    n = int(input())
    planets = [0] * n 
    parent = [0] * n
    for i in range(n):
        x, y, z = map(int, input().split(" "))
        planets[i] = (x, y, z)
        parent[i] = i

    roads = [] 
    for i in range(n-1):
        for j in range(i+1, n):
            x1, y1, z1 = planets[i]
            x2, y2, z2 = planets[j]
            distance = min(abs(x2-x1), abs(y2-y1), abs(z2-z1))
            roads.append((distance, i, j))
    roads.sort()
    result = 0
    for path in roads:
        distance, i, j = path 
        if find_p(parent, i) != find_p(parent, j):
            union(parent, i, j)
            result += distance
    return result
